tasks/vendoring/__init__.py

```diff
@@ -137,6 +155,106 @@ def vendor(ctx, vendor_dir):
         apply_patch(ctx, patch)
 
 
+def download_licenses(ctx, vendor_dir):
+    log('Downloading licenses')
+    tmp_dir = vendor_dir / '__tmp__'
+    ctx.run(
+        'pip download -r {0}/vendor.txt --no-binary :all: -d {1}'.format(
+            str(vendor_dir),
+            str(tmp_dir),
+        )
+    )
+    for sdist in tmp_dir.iterdir():
+        extract_license(vendor_dir, sdist)
+    drop_dir(tmp_dir)
+
+
+def extract_license(vendor_dir, sdist):
+    if sdist.suffixes[-2] == '.tar':
+        ext = sdist.suffixes[-1][1:]
+        with tarfile.open(sdist, mode='r:{}'.format(ext)) as tar:
+            found = find_and_extract_license(vendor_dir, tar, tar.getmembers())
+    elif sdist.suffixes[-1] == '.zip':
+        with zipfile.ZipFile(sdist) as zip:
+            found = find_and_extract_license(vendor_dir, zip, zip.infolist())
+    else:
+        raise NotImplementedError('new sdist type!')
+
+    if not found:
+        log('License not found in {}, will download'.format(sdist.name))
+        license_fallback(vendor_dir, sdist.name)
+
+
+def find_and_extract_license(vendor_dir, tar, members):
+    found = False
+    for member in members:
+        try:
+            name = member.name
+        except AttributeError:  # zipfile
+            name = member.filename
+        if 'LICENSE' in name or 'COPYING' in name:
+            if '/test' in name:
+                # some testing licenses in html5lib and distlib
+                log('Ignoring {}'.format(name))
+                continue
+            found = True
+            extract_license_member(vendor_dir, tar, member, name)
+    return found
+
+
+def license_fallback(vendor_dir, sdist_name):
+    """Hardcoded license URLs. Check when updating if those are still needed"""
+    libname = libname_from_dir(sdist_name)
+    if libname not in HARDCODED_LICENSE_URLS:
+        raise ValueError('No hardcoded URL for {} license'.format(libname))
+
+    url = HARDCODED_LICENSE_URLS[libname]
+    _, _, name = url.rpartition('/')
+    dest = license_destination(vendor_dir, libname, name)
+    r = requests.get(url, allow_redirects=True)
+    log('Downloading {}'.format(url))
+    r.raise_for_status()
+    dest.write_bytes(r.content)
+
+
+def libname_from_dir(dirname):
+    """Reconstruct the library name without it's version"""
+    parts = []
+    for part in dirname.split('-'):
+        if part[0].isdigit():
+            break
+        parts.append(part)
+    return'-'.join(parts)
```

Comment:
A space after return.

Huh. CI should catch stuff like this! Anyways, that's something I'll fix later.

Reply:
it should, right? but it's still orange.

Reply:
ee66bfd